### PR TITLE
Make parameter formatting more generic

### LINKED ISSUE #93: Can we add support for writing dictionaries as query string parameters
Hi @jonnii , it would be good to support dictionary like query string parameters, something like this:

``` c#
var fields = new Dictionary<object, object>()
{
  { "foo": "foo-values" },
  { "bar": "bar-values" }
}

httpClient.Get("authors", new { fields })
// -> translates to "GET ...authors?fields[foo]=foo-values&fields[bar]=bar-values"
```

The `ApiController` accepts such query string parameters natively (`[FromQuery] Dictionary<string, string> fields`), it would be great if speakeasy would support creating the parameters from a dictionary also

### PR DESCRIPTION
**Pull request by jonnii**
@robertcoltheart @tm011064 This extends the array formatter to cover formatting all the parameters at once. This is a breaking change, but I think perhaps this is the easiest/quickest way to close the requirement. I think it's unlikely you'll want to customize _just_ the array formatting if you're going to go down the route of customizing the way query params are generated. 

Closes #93

### CODE REVIEW

**Comment by jonnii** [comment]
ping @robertcoltheart @tm011064

**Comment by jonnii** [comment]
I didn't update the tests or clean this up, just want to get some feedback on the design before doing so.

**Comment by robertcoltheart** [comment]
So we use Speakeasy literally everywhere and I'm always going to vote for less breaking changes so that the upgrade path is easier. That said, the design looks ok to me.

**Comment by jonnii** [comment]
Either way this is going to be breaking. Is there anything else you want to change while we're at it?

**Comment by robertcoltheart** [comment]
> Is there anything else you want to change while we're at it

I'm all good I think.

**Comment by tm011064** [comment]
@jonnii had a quick look, but I don't understand the library well enough. Just to check - one would do this:

``` c#
var settings = new HttpClientSettings
{
  ParameterFormatter = new MyCustomFormatter()
};

var client = HttpClient.Create("api", settings);
```

and then the custom formatter would be used without having to add a custom middleware, right?

**Comment by jonnii** [comment]
@tm011064 exactly.

**Comment by jonnii** [comment]
@robertcoltheart done. lmk what you think. there's definitely room for improvement in terms of perf for the query param formatter (e.g. get rid of linq) but I think it's fine for now.

File: src/SpeakEasy/HttpClientSettings.cs [hunk 1]
```diff
@@ -36,7 +35,7 @@ public HttpClientSettings()
         public IAuthenticator Authenticator { get; set; }
 
         /// <summary>
-        /// The available serialiazers
+        /// The available serializers
```
Comment (robertcoltheart) [inline src/SpeakEasy/HttpClientSettings.cs]: :trollface:

**Comment by jonnii** [comment]
@robertcoltheart @tm011064 I cut `v1.2.5-beta0001`, can you try it out and make sure it meets your requirements and if so I'll cut a proper release.

**Comment by tm011064** [comment]
@jonnii thanks!

**Comment by tm011064** [comment]
@jonnii did a quick test, works great, thanks

**Comment by jonnii** [comment]
@tm011064 cutting full release now.

**Pull request merged**